### ISSUE TITLE
Keys not retained in backups but still listed in keys menu

### ISSUE THREAD

**Issue by elliot-wood** [blinksh/blink#364]
When restoring an iOS device from a backup, all ssh private keys are lost.

They will still be listed in the keys menu, but they won’t work for authentication and attempting to copy the private key causes a crash.

My guess as to the intended behaviour goes two ways. Either:
1) private keys are retained in backup, can then be used as usual, or
2) private keys are not retained, delete them from the keys menu.

Details:
* iPhone 8 Plus, iOS 11.1b2
* Using the App Store version of Blink

**Comment by goerz** [comment]
This may be related to what happened to me in #274

**Comment by carloscabanero** [comment]
Yup, this is because the private key is stored within the "Secure Enclave", and we just keep a reference to it. This is to avoid that your private keys are actually in a backup somewhere in the cloud so it is intended. 

So the behaviour should be number 2, were private keys shouldn't be retained, with probably a check on first run of the application.

Marking as bug.

**Comment by craSH** [comment]
I believe I ran in to this from an iCloud sync vs. a backup/restore. I have Blink on my iPad which has my keys, and recently installed it on a new iPhone install. iCloud sync brought over all of my host configuration, and the reference of the private key (but as mentioned here, not the key material itself due to it being protected by the secure enclave).

It may be helpful to add a note in the iCloud sync page that private keys are not included in that. Perhaps don't sync the key information at all if it's not really useful?

**Comment by moskovich** [comment]
What is the point of the option to sync the keys to iCloud then?

**Comment by carloscabanero** [comment]
This was a fk up by yours truly, because we thought we would be able to sync them, and never removed the feature just in case. I thought we had disabled it, but apparently we didn’t. Will try to fix asap.

**Comment by juneoh** [comment]
Replicated in migrating from iPhone 6 to iPhone X, and iPad 10.5 to iPad 12.9 2018. I personally believe that keys should not be synced over the cloud, in the name of security.

**Comment by moskovich** [comment]
How about storing them in Apple's keychain then?

**Comment by nreilly** [comment]
Just had this when migrating to a new iPhone. The keys listed my SE key from the previous phone and I was wondering what voodoo was happening, but it couldn’t authenticate with it (thankfully). Deleted the reference and created a new one. Perhaps a check on launch for non-existent SE keys, notifying the user that it’s not there anymore and they might like to set up a new one is a good solution.

**Comment by carloscabanero** [comment]
It took a while but just implemented a warning for the iCloud backup scenario. This should now prompt the user to recreate the keys, and we also keep the Public key for reference so keys can be recycled.

We are still thinking on ways to sync keys safely, but that will be a different feature.
